### DISCUSSION TITLE
Trouble building Adafruit Trinket M0 (and new QT PY)

### DISCUSSION THREAD

**Discussion by charkster** [micropython/micropython#14124] [SAMD] [answered]
I have not been able to get MicroPython working on my Adafruit Trinket M0. It uses a SAMD21**E**18A part (different to the SEEED Xiao's SAMD21**G**18A). I can compile and program my SEEED Xiao with no problems using the firmware.uf2 file. However after compiling the Trinket board and programing it with the UF2 file I just see error messages in dmesg. My end goal is to create a new set of board files for the Adafruit QT PY, as I have more of those boards. I am not using the Xiao UF2 on the Trinket board.. I am using the UF2 file in it's board directory.

Adafruit Trinket M0 does not have a hardware crystal oscillator... it should be configured for an on-die oscillator... I am not sure if that is missing from the mpconfigboard.h file.

I am using this command in the micropython/ports/samd directory:
`make BOARD=ADAFRUIT_TRINKET_M0`

I upgraded my Trinket UF2 boot loader to the latest version V3.7.0 from the Adafruit webpage and then tried to program the part with the created firmware.uf2 file. It appears to program, but I get the following error messages in dmesg.

```
[ 8953.561757] device offline error, dev sda, sector 127 op 0x1:(WRITE) flags 0x100000 phys_seg 1 prio class 2
[ 8953.561773] Buffer I/O error on dev sda, logical block 127, lost async page write
[ 8953.574611] FAT-fs (sda): unable to read boot sector to mark fs as dirty
[ 8955.053450] usb 3-2: new full-speed USB device number 29 using xhci_hcd
[ 8955.189494] usb 3-2: device descriptor read/64, error -71
[ 8955.433388] usb 3-2: device descriptor read/64, error -71
[ 8955.669452] usb 3-2: new full-speed USB device number 30 using xhci_hcd
[ 8955.805491] usb 3-2: device descriptor read/64, error -71
[ 8956.045256] usb 3-2: device descriptor read/64, error -71
[ 8956.157569] usb usb3-port2: attempt power cycle
[ 8956.569204] usb 3-2: new full-speed USB device number 31 using xhci_hcd
[ 8956.569327] usb 3-2: Device not responding to setup address.
[ 8956.777299] usb 3-2: Device not responding to setup address.
[ 8956.985382] usb 3-2: device not accepting address 31, error -71
[ 8957.113196] usb 3-2: new full-speed USB device number 32 using xhci_hcd
[ 8957.113304] usb 3-2: Device not responding to setup address.
[ 8957.325220] usb 3-2: Device not responding to setup address.
[ 8957.533611] usb 3-2: device not accepting address 32, error -71
[ 8957.533907] usb usb3-port2: unable to enumerate USB device
```
 
I can program different UF2 files which were created with the starting address 0x2000 (which is the same for MicroPython's firmware.uf2 file). I believe the boot loader is working properly. 

Here is the output from the MicroPython compilation: 
```
Memory region         Used Size  Region Size  %age Used
           FLASH:      184020 B       184 KB     97.67%
             RAM:        2304 B        32 KB      7.03%
Converted to uf2, output size: 368128, start address: 0x2000
Wrote 368128 bytes to build-ADAFRUIT_TRINKET_M0/firmware.uf2
```

I am confident that my Trinket board is working properly as I can get other UF2 files programmed and they properly use the USB port.

Any help would be appreciated.
Thanks.

**Comment by charkster** [comment, marked as answer]
Yep, ADAFRUIT_TRINKET_M0 mpconfigboard.h file was missing:
`#define MICROPY_HW_DFLL_USB_SYNC    (1)`

This also fixed my Adafruit QT PY board (in development).

**Comment by robert-hh** [comment]
Thank you for the analysis. The behavior is unexpected. Clearly the line setting MICROPY_HW_DFLL_USB_SYNC is missing, but the board should start without that, especially for situations where the board is not conneceted to USB. I have to dig into the clock configuration again.

**Comment by robert-hh** [comment]
It appears that the recent support for Python coded USB drivers affects the clock generation on SAMD board without a crystal. These try to use the USB SOF signal as reference. But if after a while not USB connection is established, it falls back to an internal clock reference. This is as well the state when MICROPY_HW_DFLL_USB_SYNC is not set. It turns out that on SAMD21 board the clock is then far off the proper value, and USB fails.
I prepared a PR to fix it.
